### PR TITLE
merged-fitting

### PR DESCRIPTION
**Pull request by mdada1**
Initial changes

### CODE REVIEW

**Comment by aarmey** [comment]
@mdada1 did you want me to look at something?

**Comment by mdada1** [comment]
> @mdada1 did you want me to look at something?

No, just opened this so we can have all the changes and tests in one place. I'll let you know if any issues come up!

**Comment by mdada1** [comment]
This should have the updated versions of runOptim, cost, and reshapeParams. The code I used to calculate the costs is the mergedFitting function in fitting.py.

File: de/importData.py [hunk 1]
```diff
@@ -11,6 +11,8 @@ def importLINCS(cellLine):
     data = np.load(join(path_here, "de/data/", cellLine + "_RNAi_matrix.npy"))
     annotation = pd.read_csv(join(path_here, "de/data/", cellLine + "_genes.txt"), header=None)
 
+    if data.shape[0] != data.shape[1]:
```
Comment (aarmey) [inline de/importData.py:14]: You shouldn't have to do this, so long as the control is the last column. In that case it should be handled correctly.

File: de/fitting.py [hunk 2]
```diff
@@ -1,46 +1,58 @@
...
 from scipy.optimize import minimize
-from .factorization import alpha, factorizeEstimate
+from .factorization import alpha, factorizeEstimate, cellLineComparison
+from .importData import importLINCS
 
 config.update("jax_enable_x64", True)
 
 
-def reshapeParams(p, nGenes):
+def reshapeParams(p, nGenes, nCellLines=1):
```
Comment (aarmey) [inline de/fitting.py:16]: I wonder if you can figure out nCellLines from the length of p? In general it's best to encode information in exactly one place.

**Review by aarmey — CHANGES_REQUESTED** [review]
This all looks reasonable to me. Just two questions.

File: de/factorization.py [hunk 3]
```diff
@@ -12,14 +12,29 @@
...
-    return np.linalg.lstsq(U.T, B.T, rcond=None)[0].T
+    w = None
+
+    
+    if (all(isinstance(i, np.ndarray) for i in data)):
+      for x in data:
+        U1 = np.copy(x)
+        np.fill_diagonal(U1, 0.0)
+        B = (x * alphaIn) / eta[:, np.newaxis]
+        assert np.all(np.isfinite(B))
```
Comment (Farnazmdi) [inline de/factorization.py:24]: in line 23, you are using `x` but I guess you should have `U1` since you are filling the diagonal of `U1` with zeros...

Comment (Farnazmdi) [reply de/factorization.py:24]: Try that let me know if you still have the problem.

Comment (aryakrekhi) [reply de/factorization.py:24]: The problem still exists.

**Comment by mdada1** [comment]
@aarmey Do you mind checking the changes we made to factorizeEstimate? I wasn't sure how to use both cell lines in the iterative solving process. This branch should be done otherwise.

**Review by aarmey — CHANGES_REQUESTED** [review]
Made one fix to deal with the fact that each cell line should have its own `eta`.

Please replace the current test with the following: Use the melanoma dataset, and call `factorizeEstimate` with it as one cell line. Next, duplicate the dataset to make "two" cell lines that are identical. Calculate the difference norm of `w` and `eta` in these two situations. This will be a more stringent test for the factorization model and run much faster. From there we can come up with something for the fitting code.

**Pull request merged**